### PR TITLE
Parallel network

### PR DESCRIPTION
**Pull request by emailweixu**

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -152,6 +152,110 @@ def bias(self):
...
+        n, k, l = self._weight.shape
+        if inputs.ndim == 2:
+            assert inputs.shape[1] == l, (
+                "inputs has wrong shape %s. Expecting (B, %d)" % (inputs.shape,
+                                                                  l))
+            inputs = inputs.unsqueeze(0).expand(n, *inputs.shape)
+        elif inputs.ndim == 3:
+            assert (inputs.shape[1] == n and inputs.shape[2] == l), (
+                "inputs has wrong shape %s. Expecting (B, %d, %d)" %
+                (inputs.shape, n, k))
```
Comment (Haichao-Zhang) [inline alf/layers.py]: ```(inputs.shape, n, k)``` -> ```(inputs.shape, n, l)```?

File: alf/layers.py [hunk 2]
```diff
@@ -152,6 +152,110 @@ def bias(self):
...
+        else:
+            raise ValueError("Wrong inputs.ndim=%d" % inputs.ndim)
+
+        if self.bias is not None:
+            y = torch.baddbmm(
+                self._bias.unsqueeze(1), inputs,
+                self.weight.transpose(1, 2))  # [n, B, k]
+        else:
+            y = torch.bmm(inputs, self._weight.transpose(1, 2))  # [n, B, k]
+        y = y.transpose(0, 1)  # [B, n, l]
```
Comment (Haichao-Zhang) [inline alf/layers.py]: ```[B, n, l]```->```[B, n, k]```

File: alf/networks/critic_networks.py [hunk 3]
```diff
@@ -143,6 +145,56 @@ def forward(self, inputs, state=()):
...
+                 critic_network: CriticNetwork,
+                 n: int,
+                 name='ParallelCriticNetwork'):
+        """
+        It create a parallelized version of ``critic_network``.
+        
+        Args:
+            critic_network (CriticNetwork): non-parallelized critic network
+            n (int): make ``n`` replicas from ``critic_network`` with different
+                initializaztion.
```
Comment (Haichao-Zhang) [inline alf/networks/critic_networks.py]: ```initializaztion``` -> ```initialization```

File: alf/networks/encoding_networks.py [hunk 4]
```diff
@@ -403,13 +405,155 @@ def forward(self, inputs, state=()):
...
-        if self._output_spec is None:
-            self._output_spec = TensorSpec(
-                (self._output_size, ),
-                dtype=self._processed_input_tensor_spec.dtype)
-        return self._output_spec
+    def make_parallel(self, n):
+        """Make a parllelized version of this network.
+        
+        A parallel network has ``n`` copies of network with the same structure but
+        different indepently initialized parameters.
```
Comment (Haichao-Zhang) [inline alf/networks/encoding_networks.py]: ```indepently``` -> ```independently```

File: alf/layers.py (hunk 1 above)
Comment (emailweixu) [reply alf/layers.py]: fixed.

File: alf/layers.py (hunk 2 above)
Comment (emailweixu) [reply alf/layers.py]: fixed

File: alf/networks/critic_networks.py (hunk 3 above)
Comment (emailweixu) [reply alf/networks/critic_networks.py]: fixed.

**Review by Haichao-Zhang — APPROVED** [review]

**Pull request merged**